src/main/java/groupM/players/mcts/TreeNode.java

```diff
@@ -212,25 +257,32 @@ private boolean finishRollout(AbstractGameState rollerState, int depth) {
         return !rollerState.isNotTerminal();
     }
 
-    /**
-     * Selects the action to take from the current node.
-     * Must handle the explore/exploit dilemma
-     * @return the action
-     */
-    abstract AbstractAction selectAction();
+    private AbstractAction selectAction() {
+     AbstractAction bestAction = null;
+     double bestValue = -Double.MAX_VALUE;
 
-    /**
-     * Back up the value of the child through all parents.
-     *
-     * @param result - value of rollout to backup
-     */
-    abstract void backUp(double result);
+     for (AbstractAction action : children.keySet()) {
+        UCB1TreeNode child = (UCB1TreeNode) children.get(action);
```

Comment:
What's the relationship between TreeNode and UCB1TreeNode? Should TreeNode assume it's children are of type UCBTreeNode?